### PR TITLE
Make the progresscontainer independent of each other

### PR DESCRIPTION
**Pull request by richardpen**
### Summary

Change the waitForContainerTransitions to only wait for one transition instead of all of the transitions so that each container can still move forward on its own without waiting for other transitions.

### Implementation details

Remove the logic of waiting for all transitions in waitForContainerTransitions, and added a transition status to indicate the container is in the middle of a processing to avoid duplicate action on the same container.

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass
- [x] Manual test with task that has multiple containers, containers are progressed independently.
  * Task with multiple containers
  * Task that uses volume (including empty volume and non-empty volume)
  * Task that uses volumesFrom
  * Task that uses links
  * Set a cluster to periodically run a bunch of tasks

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/api/container.go [hunk 1]
```diff
@@ -548,3 +553,38 @@ func (c *Container) GetHealthStatus() HealthStatus {
...
+
+// UpdateTransitioningStatusUnsafe updates the container transitioning status
+func (c *Container) UpdateTransitioningStatusUnsafe(knownStatus ContainerStatus) {
+	if c.transitioningStatus == ContainerNotTransition {
+		return
+	}
+
+	// Check if the container transition has already finished
+	if containerTransitioningStatusMapping[c.transitioningStatus] <= knownStatus {
+		c.transitioningStatus = ContainerNotTransition
```
Comment (adnxn) [inline agent/api/container.go]: do we need a lock for this? also should we just be using the setter here?

**Comment by richardpen** [comment]
@adnxn For question 1: functional tests only test the happy path, I will need to test the failure case where a container failed at different stage:
* Task with multiple containers
* Task that uses volume (including empty volume and non-empty volume)
* Task that uses volumesFrom
* Task that uses links
* Set a cluster to periodically run a bunch of tasks

Let me know if you can think of other test scenario that was missed here.

For 2, we don't need to save the state, as agent has already has the logic to handle duplicate create/start, for pull/stop agent will have to pull/stop again on agent restart.

File: agent/api/container.go (hunk 1 above)
Comment (richardpen) [reply agent/api/container.go]: This method should be called within SetKnownStatus, which has already held the lock. Setter is holding a lock where we don't need a lock here, that's why the name xxxUnsafe.

File: agent/api/container.go [hunk 2]
```diff
@@ -548,3 +553,38 @@ func (c *Container) GetHealthStatus() HealthStatus {
...
+		// return false to indicate the set operation failed
+		return false
+	}
+
+	c.transitioningStatus = status
+	return true
+}
+
+// GetTransitioning returns the transitioning status of container
+func (c *Container) GetTransitioning() ContainerTransitioningStatus {
```
Comment (aaithal) [inline agent/api/container.go]: can you please rename this to `TransitionStatus()`?

File: agent/api/containerstatus.go [hunk 3]
```diff
@@ -49,9 +49,28 @@ const (
 	ContainerUnhealthy
 )
 
+const (
+	// ContainerNotTransition means the container isn't in the middle of a transition
+	ContainerNotTransition ContainerTransitioningStatus = iota
```
Comment (aaithal) [inline agent/api/containerstatus.go]: Can this be `ContainerTransitionNone`?

File: agent/api/containerstatus.go [hunk 4]
```diff
@@ -49,9 +49,28 @@ const (
...
+	// ContainerPulling means the container is in the process of pulling
+	ContainerPulling
```
Comment (aaithal) [inline agent/api/containerstatus.go]: Can you make sure that `ContainerTransition` is a prefix for all of these? Example: `ContainerTransitionPulling`

File: agent/api/containerstatus.go [hunk 5]
```diff
@@ -49,9 +49,28 @@ const (
...
+	// ContainerStopping means the container is in the process of stopping
+	ContainerStopping
+)
+
 // ContainerStatus is an enumeration of valid states in the container lifecycle
 type ContainerStatus int32
 
+// ContainerTransitioningStatus is an enumeration of valid container processing
+// status which indicates which status the container is being processed
+type ContainerTransitioningStatus int32
```
Comment (aaithal) [inline agent/api/containerstatus.go]: Can you move all of Transition statuses and methods to a separate file?

File: agent/api/container.go [hunk 6]
```diff
@@ -548,3 +553,38 @@ func (c *Container) GetHealthStatus() HealthStatus {
...
+
+	// Check if the container transition has already finished
+	if containerTransitioningStatusMapping[c.transitioningStatus] <= knownStatus {
+		c.transitioningStatus = ContainerNotTransition
+	}
+}
+
+// SetTransitioningStatus sets the transitioning status of container and returns whether
+// the container is already in a transition
+func (c *Container) SetTransitioningStatus(status ContainerTransitioningStatus) bool {
```
Comment (aaithal) [inline agent/api/container.go]: `UpdateTransitioningTo` might be a better name here.

File: agent/engine/task_manager.go [hunk 7]
```diff
@@ -663,16 +663,26 @@ func (mtask *managedTask) startContainerTransitions(transitionFunc containerTran
...
 		anyCanTransition = true
 
 		if !transition.actionRequired {
-			mtask.handleContainerChange(dockerContainerChange{
-				container: cont,
-				event: DockerContainerChangeEvent{
-					Status: transition.nextState,
-				},
-			})
+			go func(cont *api.Container, status api.ContainerStatus) {
```
Comment (aaithal) [inline agent/engine/task_manager.go]: why was this logic changed? Can you please add documentation/comments for that as well?

File: agent/engine/task_manager.go [hunk 8]
```diff
@@ -763,39 +773,21 @@ func (mtask *managedTask) onContainersUnableToTransitionState() {
...
-			if allWaitingOnPulled {
-				// We don't actually care to wait for 'pull' transitions to finish if
-				// we're just heading to stopped since those resources aren't
-				// inherently linked to this task anyways for e.g. gc and so on.
-				seelog.Debugf("Managed task [%s]: all containers are waiting for pulled transition; exiting early: %v",
-					mtask.Arn, transitions)
-				break
-			}
-		}
+	// Wait for one transition/acs/docker message
```
Comment (aaithal) [inline agent/engine/task_manager.go]: Same comment here. Why was this changed? What's the impact? Can we please add documentation explaining the change?

File: agent/api/container.go [hunk 9]
```diff
@@ -166,6 +166,9 @@ type Container struct {
 	// handled properly so that the state storage continues to work.
 	SentStatusUnsafe ContainerStatus `json:"SentStatus"`
 
+	// transitioningStatus is the status of the container
+	transitioningStatus ContainerTransitioningStatus
```
[the remark this answers is not in the document]
Comment (richardpen) [reply agent/api/container.go]: Thanks for pointing this out, I don't know there is already `AppliedStatus` for the same purpose. After some investigation, I think we can just use the `AppliedStatus` and the Transitioning status related should be removed. I will update the PR.

File: agent/engine/task_manager.go [hunk 10]
```diff
@@ -663,16 +663,29 @@ func (mtask *managedTask) startContainerTransitions(transitionFunc containerTran
...
-				container: cont,
-				event: DockerContainerChangeEvent{
-					Status: transition.nextState,
-				},
-			})
+			// Updating the container status without calling any docker API, send in
+			// a goroutine so that it won't block here before the waitForContainerTransitions
+			// was called after this function.
+			go func(cont *api.Container, status api.ContainerStatus) {
+				mtask.dockerMessages <- dockerContainerChange{
```
Comment (sharanyad) [inline agent/engine/task_manager.go]: For my understanding - how do we still handle the container status changes here that `handleContainerChange` does when no action is required for the transition ?

File: agent/api/container.go [hunk 11]
```diff
@@ -548,3 +552,38 @@ func (c *Container) GetHealthStatus() HealthStatus {
...
 }
+
+// updateAppliedStatusUnsafe updates the container transitioning status
+func (c *Container) updateAppliedStatusUnsafe(knownStatus ContainerStatus) {
+	if c.AppliedStatus == ContainerStatusNone {
+		return
+	}
+
+	// Check if the container transition has already finished
+	if c.AppliedStatus <= knownStatus {
```
[the remark this answers is not in the document]
Comment (richardpen) [reply agent/api/container.go]: If the agent received duplicate docker events for some reason, this could happen. As the event can be a past status of the container.

File: agent/engine/task_manager.go [hunk 12]
```diff
@@ -763,39 +776,23 @@ func (mtask *managedTask) onContainersUnableToTransitionState() {
 		mtask.emitTaskEvent(mtask.Task, taskUnableToTransitionToStoppedReason)
 		// TODO we should probably panic here
 	} else {
-		seelog.Criticalf("Managed task [%s]: voving task to stopped due to bad state", mtask.Arn)
+		seelog.Criticalf("Managed task [%s]: moving task to stopped due to bad state", mtask.Arn)
 		mtask.handleDesiredStatusChange(api.TaskStopped, 0)
 	}
 }
 
 func (mtask *managedTask) waitForContainerTransitions(transitions map[string]api.ContainerStatus,
```
[the remark this answers is not in the document]
Comment (richardpen) [reply agent/engine/task_manager.go]: But it's still waiting for container transition. Let me know if you have a better suggestion.

File: agent/engine/task_manager.go (hunk 10 above)
Comment (richardpen) [reply agent/engine/task_manager.go]: All the events sent to `mtask.dockerMessage` will be handled by `handleContainerChange` which works the same as before except that we don't call `handleContainerChange` directly now. Is that what you are asking?

**Comment by aaithal** [comment]
This requires a CHANGELOG entry as well. Also, should this be tagged for 1.17.3 milestone?

File: agent/engine/task_manager.go (hunk 10 above)
Comment (aaithal) [reply agent/engine/task_manager.go]: >All the events sent to `mtask.dockerMessage` will be handled by `handleContainerChange` which works the same as before except that we don't call `handleContainerChange` directly now

It'd be awesome if you added that as code comment

File: agent/statemanager/state_manager.go [hunk 13]
```diff
@@ -60,7 +60,8 @@ const (
 	// 9) Add 'ipToTask' map to state file
 	// 10) Add 'healthCheckType' field in 'api.Container'
 	// 11) Add 'PrivateDNSName' field to 'api.ENI'
-	ECSDataVersion = 11
+	// 12) Remove `AppliedStatus` field form 'api.Container'
+	ECSDataVersion = 12
```
Comment (aaithal) [inline agent/statemanager/state_manager.go]: This should still be `11`. `10` is what's released out there.

File: agent/engine/task_manager.go (hunk 10 above)
Comment (sharanyad) [reply agent/engine/task_manager.go]: @richardpen Got it, thanks

File: agent/engine/task_manager.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (sharanyad) [reply agent/engine/task_manager.go]: oh it just makes it look like it is waiting for ALL the containers' transitions, while it is waiting for just one to finish. may be just call it `waitForContainerTransition`?

File: agent/engine/task_manager.go [hunk 14]
```diff
@@ -763,39 +777,23 @@ func (mtask *managedTask) onContainersUnableToTransitionState() {
...
-				// inherently linked to this task anyways for e.g. gc and so on.
-				seelog.Debugf("Managed task [%s]: all containers are waiting for pulled transition; exiting early: %v",
-					mtask.Arn, transitions)
-				break
-			}
-		}
+	// There could be multiple transitions, but we just need to wait for one of them
+	// to ensure that there is at least one container can be processed in the next
+	// progressContainers. This is done by waiting for one transition/acs/docker message.
+	if mtask.waitEvent(transitionChange) {
```
Comment (sharanyad) [inline agent/engine/task_manager.go]: If we do not wait for a container transition here, next call of `progressContainers` would happen, which may be a no-op. is that the reason we are waiting here?

Comment (richardpen) [reply agent/engine/task_manager.go]: yes, otherwise it will exhaust the cpu resource with an empty for loop.

File: agent/api/container.go [hunk 15]
```diff
@@ -548,3 +552,38 @@ func (c *Container) GetHealthStatus() HealthStatus {
...
+}
+
+// SetAppliedStatus sets the applied status of container and returns whether
+// the container is already in a transition
+func (c *Container) SetAppliedStatus(status ContainerStatus) bool {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.AppliedStatus != ContainerStatusNone {
+		// return false to indicate the set operation failed
```
[the remark this answers is not in the document]
Comment (richardpen) [reply agent/api/container.go]: No, failed means the container is already in a transition(where the appliedstatus isn't none). This ensures that agent won't call the same API(pull/create/start/stop) twice for the same container.

File: agent/engine/task_manager.go [hunk 16]
```diff
@@ -763,39 +777,23 @@ func (mtask *managedTask) onContainersUnableToTransitionState() {
...
-			seelog.Debugf("Managed task [%s]: still waiting for: %v", mtask.Arn, transitions)
-		}
-		if mtask.GetDesiredStatus().Terminal() || mtask.GetKnownStatus().Terminal() {
-			allWaitingOnPulled := true
-			for _, desired := range transitions {
-				if desired != api.ContainerPulled {
-					allWaitingOnPulled = false
-				}
-			}
-			if allWaitingOnPulled {
```
Comment (sharanyad) [inline agent/engine/task_manager.go]: How is this condition handled in this new change?

Comment (richardpen) [reply agent/engine/task_manager.go]: In previously the pull will block the whole transitions, that's why we want to break the wait if allWaitingOnPulled. Where the new change makes it non-blocked, which is handled by default.

File: agent/engine/task_manager.go [hunk 17]
```diff
@@ -763,39 +777,23 @@ func (mtask *managedTask) onContainersUnableToTransitionState() {
...
-				seelog.Debugf("Managed task [%s]: all containers are waiting for pulled transition; exiting early: %v",
-					mtask.Arn, transitions)
-				break
-			}
-		}
+	// There could be multiple transitions, but we just need to wait for one of them
+	// to ensure that there is at least one container can be processed in the next
+	// progressContainers. This is done by waiting for one transition/acs/docker message.
+	if mtask.waitEvent(transitionChange) {
+		changedContainer := <-transitionChangeContainer
```
Comment (aaithal) [inline agent/engine/task_manager.go]: minor: `transitionedContainer` is a better name for this

File: agent/engine/task_manager.go [hunk 18]
```diff
@@ -763,39 +777,23 @@ func (mtask *managedTask) onContainersUnableToTransitionState() {
...
-				// inherently linked to this task anyways for e.g. gc and so on.
-				seelog.Debugf("Managed task [%s]: all containers are waiting for pulled transition; exiting early: %v",
-					mtask.Arn, transitions)
-				break
-			}
-		}
+	// There could be multiple transitions, but we just need to wait for one of them
+	// to ensure that there is at least one container can be processed in the next
+	// progressContainers. This is done by waiting for one transition/acs/docker message.
+	if mtask.waitEvent(transitionChange) {
```
Comment (aaithal) [inline agent/engine/task_manager.go]: minor: `transition` is a better name for this

File: agent/engine/task_manager.go [hunk 19]
```diff
@@ -618,8 +618,8 @@ func (mtask *managedTask) progressContainers() {
 	// We've kicked off one or more transitions, wait for them to
 	// complete, but keep reading events as we do. in fact, we have to for
 	// transitions to complete
-	mtask.waitForContainerTransitions(transitions, transitionChange, transitionChangeContainer)
-	seelog.Debugf("Managed task [%s]: done transitioning all containers", mtask.Arn)
+	mtask.waitForContainerTransition(transitions, transitionChange, transitionChangeContainer)
+	seelog.Debugf("Managed task [%s]: wait for container transition done", mtask.Arn)
```
Comment (aaithal) [inline agent/engine/task_manager.go]: I don't think we need this line. You're logging in `waitForContainerTransition()` anyway

File: agent/engine/task_manager.go (hunk 18 above)
Comment (aaithal) [inline agent/engine/task_manager.go]: This also needs an `else` part, where we can log that we were interrupted or the transition did not finish or something like that.
